conda_smithy/feedstock_tokens.py

```diff
@@ -0,0 +1,586 @@
+            pth = os.path.join(
+                "~",
+                ".conda-smithy",
+                "%s_%s_feedstock.token" % (user, project),
+            )
+            pth = os.path.expanduser(pth)
+            if os.path.exists(pth):
+                failed = True
+                err_msg = "Token for %s%s is already written locally!" % (
+                    user,
+                    project,
+                )
+                raise RuntimeError(err_msg)
+
+            os.makedirs(os.path.dirname(pth), exist_ok=True)
+
+            with open(pth, "w") as fp:
+                fp.write(token)
+        except Exception as e:
+            if "DEBUG_FEEDSTOCK_TOKENS" in os.environ:
+                raise e
+            failed = True
+
+    if failed:
+        if err_msg:
+            raise RuntimeError(err_msg)
+        else:
+            raise RuntimeError(
+                (
+                    "Generating the feedstock token for %s/%s failed!"
+                    " Try the command locally with DEBUG_FEEDSTOCK_TOKENS"
+                    " defined in the environment to investigate!"
+                )
+                % (user, project)
+            )
+
+    return failed
+
+
+def read_feedstock_token(user, project):
+    """Read the feedstock token from
+
+        ~/.conda-smithy/{user or org}_{repo w/o '-feedstock'}_feedstock.token
+
+    In order to not spill any tokens to stdout/stderr, this function
+    should be used in a `try...except` block with stdout/stderr redirected
+    to /dev/null, etc.
+    """
+    err_msg = None
+    feedstock_token = None
+
+    # read the token
+    user_token_pth = os.path.join(
+        "~", ".conda-smithy", "%s_%s_feedstock.token" % (user, project),
+    )
+    user_token_pth = os.path.expanduser(user_token_pth)
+
+    if not os.path.exists(user_token_pth):
+        err_msg = (
+            "No token found in '~/.conda-smithy/%s_%s_feedstock.token'"
+            % (user, project,)
+        )
+    else:
+        with open(user_token_pth, "r") as fp:
+            feedstock_token = fp.read().strip()
+        if not feedstock_token:
+            err_msg = (
+                "Empty token found in '~/.conda-smithy/"
+                "%s_%s_feedstock.token'"
+            ) % (user, project,)
+    return feedstock_token, err_msg
+
+
+def feedstock_token_exists(user, project, token_repo):
+    """Test if the feedstock token exists for the given repo.
+
+    All exceptions are swallowed and stdout/stderr from this function is
+    redirected to `/dev/null`. Sanitized error messages are
+    displayed at the end.
+
+    If you need to debug this function, define `DEBUG_FEEDSTOCK_TOKENS` in
+    your environment before calling this function.
+    """
+    from .github import gh_token
+
+    github_token = gh_token()
+
+    exists = False
+    failed = False
+    with tempfile.TemporaryDirectory() as tmpdir:
+        with open(os.devnull, "w") as fp, redirect_stdout(fp), redirect_stderr(
```

Comment:
Actually this can be combined with the above as well too, if we want to reduce the indentation even more

Reply:
indeed it can. I'll push a change later